File: prisma/seed.mjs
import { PrismaClient } from "@prisma/client";
import { createHash } from "crypto";

const prisma = new PrismaClient();
const H = (p) => createHash("sha256").update(`dokkanek:${p}`).digest("hex");
const daysAgo = (n) => new Date(Date.now() - n * 86400000);

async function main() {
  // الإعدادات
  const settings = {
    store_name: "دكّانك",
    logo_text: "دكّانك",
    primary_color: "#0d6efd",
    phone: "0910000000",
    address: "طرابلس — ليبيا",
    footer_note: "شكرا لتسوقكم معنا",
    currency: "د.ل",
  };
  for (const [key, value] of Object.entries(settings)) {
    await prisma.setting.upsert({ where: { key }, update: { value }, create: { key, value } });
  }

  // الفروع والمخازن
  const b1 = await prisma.branch.upsert({
    where: { id: "seed-branch-tripoli" },
    update: {},
    create: { id: "seed-branch-tripoli", name: "الفرع الرئيسي", city: "طرابلس", phone: "0910000001" },
  });
  const b2 = await prisma.branch.upsert({
    where: { id: "seed-branch-benghazi" },
    update: {},
    create: { id: "seed-branch-benghazi", name: "فرع بنغازي", city: "بنغازي", phone: "0920000002" },
  });
  const w1 = await prisma.warehouse.upsert({
    where: { id: "seed-wh-main" },
    update: {},
    create: { id: "seed-wh-main", name: "المخزن الرئيسي", branchId: b1.id },
  });

  // المستخدمون (كلمات المرور للتجربة فقط)
  const users = [
    { username: "admin", name: "مدير النظام", role: "ADMIN", password: "admin123", branchId: b1.id },
    { username: "cashier", name: "أحمد الكاشير", role: "CASHIER", password: "1234", branchId: b1.id },
    { username: "courier", name: "سالم المندوب", role: "COURIER", password: "1234", branchId: b1.id },
    { username: "tech", name: "مفتاح الفني", role: "TECHNICIAN", password: "1234", branchId: b1.id },
  ];
  for (const u of users) {
    await prisma.user.upsert({
      where: { username: u.username },
      update: { passwordHash: H(u.password), role: u.role, name: u.name, active: true },
      create: { username: u.username, name: u.name, role: u.role, passwordHash: H(u.password), branchId: u.branchId },
    });
  }

  // التصنيفات
  const catNames = ["هواتف", "إكسسوارات", "قطع غيار صيانة", "أجهزة لوحية"];
  const cats = {};
  for (const n of catNames) {
    const c = await prisma.category.upsert({ where: { name: n }, update: {}, create: { name: n } });
    cats[n] = c.id;
  }

  // أصناف وهمية (أسماء عامة بدون علامات تجارية)
  const products = [
    { sku: "PH-001", name: "هاتف ذكي فئة اقتصادية 128GB", cat: "هواتف", cost: 780, price: 950, qty: 12, fav: true, min: 3, barcode: "100001" },
    { sku: "PH-002", name: "هاتف ذكي فئة متوسطة 256GB", cat: "هواتف", cost: 1450, price: 1720, qty: 7, fav: true, min: 2, barcode: "100002" },
    { sku: "PH-003", name: "هاتف مستعمل مجدد — حالة ممتازة", cat: "هواتف", cost: 420, price: 550, qty: 4, fav: false, min: 2, barcode: "100003" },
    { sku: "TB-001", name: "جهاز لوحي 10 بوصة 64GB", cat: "أجهزة لوحية", cost: 620, price: 790, qty: 6, fav: false, min: 2, barcode: "100004" },
    { sku: "AC-001", name: "سماعات بلوتوث لاسلكية", cat: "إكسسوارات", cost: 55, price: 95, qty: 40, fav: true, min: 10, barcode: "100005" },
    { sku: "AC-002", name: "شاحن سريع 25W مع كابل", cat: "إكسسوارات", cost: 28, price: 55, qty: 60, fav: true, min: 15, barcode: "100006" },
    { sku: "AC-003", name: "جراب حماية شفاف", cat: "إكسسوارات", cost: 6, price: 20, qty: 120, fav: false, min: 30, barcode: "100007" },
    { sku: "AC-004", name: "لاصق حماية شاشة مقوى", cat: "إكسسوارات", cost: 4, price: 15, qty: 200, fav: false, min: 50, barcode: "100008" },
    { sku: "AC-005", name: "باور بانك 20000mAh", cat: "إكسسوارات", cost: 85, price: 140, qty: 18, fav: true, min: 5, barcode: "100009" },
    { sku: "AC-006", name: "ساعة ذكية رياضية", cat: "إكسسوارات", cost: 130, price: 210, qty: 9, fav: false, min: 3, barcode: "100010" },
    { sku: "AC-007", name: "كابل شحن سريع Type-C متر ونص", cat: "إكسسوارات", cost: 9, price: 25, qty: 3, fav: false, min: 10, barcode: "100011" },
    { sku: "SP-001", name: "شاشة بديلة أصلية — فئة متوسطة", cat: "قطع غيار صيانة", cost: 160, price: 230, qty: 8, fav: false, min: 3, barcode: "100012" },
    { sku: "SP-002", name: "بطارية بديلة عالية السعة", cat: "قطع غيار صيانة", cost: 45, price: 80, qty: 15, fav: false, min: 5, barcode: "100013" },
    { sku: "SP-003", name: "قاعدة شحن وكاوية صيانة", cat: "قطع غيار صيانة", cost: 90, price: 150, qty: 2, fav: false, min: 2, barcode: "100014" },
  ];
  const prodIds = {};
  for (const p of products) {
    const r = await prisma.product.upsert({
      where: { sku: p.sku },
      update: { name: p.name, costPrice: p.cost, salePrice: p.price, quantity: p.qty, isFavorite: p.fav, minQuantity: p.min, barcode: p.barcode, categoryId: cats[p.cat], warehouseId: w1.id, active: true },
      create: { sku: p.sku, name: p.name, costPrice: p.cost, salePrice: p.price, quantity: p.qty, isFavorite: p.fav, minQuantity: p.min, barcode: p.barcode, categoryId: cats[p.cat], warehouseId: w1.id },
    });
    prodIds[p.sku] = r.id;
  }

  // زبائن وموردون
  const custNames = [
    { name: "محمد الفيتوري", phone: "0911111111", limit: 2000 },
    { name: "عائشة بن عامر", phone: "0922222222", limit: 1000 },
    { name: "خالد المصراتي", phone: "0933333333", limit: 5000 },
    { name: "فاطمة الزهراء", phone: "0944444444", limit: 500 },
    { name: "زبون نقدي", phone: "", limit: 0 },
  ];
  const custIds = [];
  for (const c of custNames) {
    const r = await prisma.customer.upsert({
      where: { id: `seed-cust-${c.phone || "cash"}` },
      update: {},
      create: { id: `seed-cust-${c.phone || "cash"}`, name: c.name, phone: c.phone, creditLimit: c.limit },
    });
    custIds.push(r.id);
  }
  const supNames = ["شركة النور للتوريدات", "مؤسسة الأفق للإلكترونيات", "مخازن الاتحاد"];
  const supIds = [];
  for (let i = 0; i < supNames.length; i++) {
    const r = await prisma.supplier.upsert({
      where: { id: `seed-sup-${i}` },
      update: {},
      create: { id: `seed-sup-${i}`, name: supNames[i], phone: `095000000${i}` },
    });
    supIds.push(r.id);
  }

  // فواتير بيع وهمية
  const cashier = await prisma.user.findUnique({ where: { username: "cashier" } });
  const saleSeeds = [
    { no: "S-DEMO-001", ago: 0, status: "COMPLETED", pay: "CASH", items: [["AC-001", 2], ["AC-003", 3]], cust: 4 },
    { no: "S-DEMO-002", ago: 0, status: "PENDING", pay: "CASH", items: [["PH-001", 1]], cust: 0 },
    { no: "S-DEMO-003", ago: 1, status: "COMPLETED", pay: "CARD", items: [["AC-005", 1], ["AC-002", 2]], cust: 4 },
    { no: "S-DEMO-004", ago: 1, status: "COURIER", pay: "CASH", items: [["AC-006", 1]], cust: 1 },
    { no: "S-DEMO-005", ago: 2, status: "COMPLETED", pay: "CREDIT", items: [["PH-002", 1]], cust: 2 },
    { no: "S-DEMO-006", ago: 3, status: "COMPLETED", pay: "CASH", items: [["AC-004", 5], ["AC-007", 2]], cust: 4 },
  ];
  for (const s of saleSeeds) {
    const exists = await prisma.sale.findUnique({ where: { no: s.no } });
    if (exists) continue;
    let subtotal = 0;
    for (const [sku, qty] of s.items) {
      const p = await prisma.product.findUnique({ where: { sku } });
      if (p) subtotal += p.salePrice * qty;
    }
    const paid = s.pay === "CREDIT" ? 0 : subtotal;
    const sale = await prisma.sale.create({
      data: {
        no: s.no, branchId: b1.id, customerId: custIds[s.cust], cashierId: cashier?.id,
        status: s.status, payMethod: s.pay, subtotal, discount: 0, total: subtotal, paid,
        date: daysAgo(s.ago),
      },
    });
    for (const [sku, qty] of s.items) {
      const p = await prisma.product.findUnique({ where: { sku } });
      if (!p) continue;
      await prisma.saleItem.create({ data: { saleId: sale.id, productId: p.id, qty, price: p.salePrice } });
      if (s.status === "COMPLETED" || s.status === "COURIER") {
        await prisma.product.update({ where: { id: p.id }, data: { quantity: Math.max(0, p.quantity - qty) } });
        await prisma.stockMove.create({ data: { productId: p.id, qty: -qty, type: "OUT", note: `فاتورة ${s.no}` } });
      }
    }
    if (s.pay !== "CREDIT") {
      await prisma.payment.create({ data: { saleId: sale.id, customerId: custIds[s.cust], amount: paid, method: s.pay } });
    } else {
      await prisma.customer.update({ where: { id: custIds[s.cust] }, data: { balance: { increment: subtotal } } });
    }
    if (s.status === "COURIER") {
      await prisma.courierTask.create({ data: { saleId: sale.id, courierName: "سالم المندوب", status: "WITH_COURIER", codAmount: subtotal } });
    }
  }

  // مصروفات وتذاكر صيانة وموظفون
  await prisma.expense.deleteMany({});
  await prisma.expense.createMany({
    data: [
      { branchId: b1.id, title: "إيجار المحل", amount: 2500, date: daysAgo(2) },
      { branchId: b1.id, title: "كهرباء وإنترنت", amount: 320, date: daysAgo(1) },
      { branchId: b1.id, title: "مواصلات المندوب", amount: 150, date: daysAgo(0) },
    ],
  });

  const tickets = [
    { no: "T-001", name: "عبد الله الحاسي", phone: "0915555555", device: "هاتف — كسر شاشة", status: "DIAGNOSIS", tech: "مفتاح الفني", cost: 180 },
    { no: "T-002", name: "مريم القذافي", phone: "0926666666", device: "هاتف — بطارية", status: "READY", tech: "مفتاح الفني", cost: 90 },
    { no: "T-003", name: "يوسف الديب", phone: "0937777777", device: "جهاز لوحي — سوكت شحن", status: "RECEIVED", tech: "", cost: 0 },
  ];
  for (const t of tickets) {
    await prisma.maintenanceTicket.upsert({
      where: { no: t.no },
      update: {},
      create: { no: t.no, customerName: t.name, customerPhone: t.phone, device: t.device, status: t.status, technician: t.tech, cost: t.cost, receivedAt: daysAgo(1) },
    });
  }

  for (const e of [
    { id: "seed-emp-1", name: "أحمد الكاشير", title: "كاشير", salary: 1500 },
    { id: "seed-emp-2", name: "سالم المندوب", title: "مندوب توصيل", salary: 1200 },
  ]) {
    await prisma.employee.upsert({ where: { id: e.id }, update: {}, create: e });
  }

  await prisma.auditLog.create({
    data: { username: "النظام", action: "SEED", entity: "Database", details: "تحميل بيانات تجريبية" },
  });

  console.log("Seed OK");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: windows/src/Dokkanek.Desktop/Data/Seed.cs
using System;
using System.Security.Cryptography;
using System.Text;
using System.Data.SQLite;

namespace Dokkanek.Desktop.Data
{
    /// <summary>
    /// First-run seed. Runs only when the User table is empty.
    /// All SQL is parameterized; password hashes match lib/auth.ts SHA256("dokkanek:"+pwd).
    /// </summary>
    public static class Seed
    {
        public static void Seed()
        {
            SeedWith(Db.ConnectionString);
        }

        public static void SeedDefault()
        {
            SeedWith(Db.ConnectionString);
        }

        public static void SeedWith(string connectionString)
        {
            if (connectionString == null)
            {
                throw new ArgumentNullException("connectionString");
            }
            Db.EnsureCreated();
            using (SQLiteConnection conn = Db.Open(connectionString))
            {
                long users;
                using (SQLiteCommand count = conn.CreateCommand())
                {
                    count.CommandText = "SELECT COUNT(*) FROM \"User\";";
                    object v = count.ExecuteScalar();
                    users = (v == null || v == DBNull.Value) ? 0 : Convert.ToInt64(v);
                }
                if (users > 0)
                {
                    return;
                }
                using (SQLiteTransaction tx = conn.BeginTransaction())
                {
                    string now = DateTime.UtcNow.ToString("o");

                    // Branches: Tripoli + Benghazi
                    Exec(conn, tx, "INSERT INTO \"Branch\"(Id, Name, City, Phone) VALUES(@id,@name,@city,@phone);",
                        P("@id", "br-tripoli"), P("@name", "Tripoli Branch"), P("@city", "Tripoli"), P("@phone", ""));
                    Exec(conn, tx, "INSERT INTO \"Branch\"(Id, Name, City, Phone) VALUES(@id,@name,@city,@phone);",
                        P("@id", "br-benghazi"), P("@name", "Benghazi Branch"), P("@city", "Benghazi"), P("@phone", ""));

                    // Warehouse
                    Exec(conn, tx, "INSERT INTO \"Warehouse\"(Id, Name, BranchId) VALUES(@id,@name,@branch);",
                        P("@id", "wh-main"), P("@name", "Main Warehouse"), P("@branch", "br-tripoli"));

                    // Users (compat hash SHA256("dokkanek:"+pwd))
                    InsertUser(conn, tx, "u-admin", "System Admin", "admin", HashPassword("admin123"), "ADMIN", "br-tripoli", now);
                    InsertUser(conn, tx, "u-cashier", "Cashier", "cashier", HashPassword("1234"), "CASHIER", "br-tripoli", now);
                    InsertUser(conn, tx, "u-courier", "Courier", "courier", HashPassword("1234"), "COURIER", "br-tripoli", now);
                    InsertUser(conn, tx, "u-tech", "Technician", "tech", HashPassword("1234"), "TECHNICIAN", "br-tripoli", now);

                    // Categories
                    Exec(conn, tx, "INSERT INTO \"Category\"(Id, Name) VALUES(@id,@name);", P("@id", "cat-phones"), P("@name", "Phones"));
                    Exec(conn, tx, "INSERT INTO \"Category\"(Id, Name) VALUES(@id,@name);", P("@id", "cat-tablets"), P("@name", "Tablets"));
                    Exec(conn, tx, "INSERT INTO \"Category\"(Id, Name) VALUES(@id,@name);", P("@id", "cat-accessories"), P("@name", "Accessories"));
                    Exec(conn, tx, "INSERT INTO \"Category\"(Id, Name) VALUES(@id,@name);", P("@id", "cat-parts"), P("@name", "Spare Parts"));

                    // Products (14): PH/TB/AC/SP prefixes
                    InsertProduct(conn, tx, "p-01", "PH-IP14", "iPhone 14 128GB", "cat-phones", 3200, 0, 3650, 12, now);
                    InsertProduct(conn, tx, "p-02", "PH-SM-A54", "Samsung Galaxy A54", "cat-phones", 1400, 0, 1650, 20, now);
                    InsertProduct(conn, tx, "p-03", "PH-RM-11", "Redmi Note 11", "cat-phones", 750, 0, 900, 30, now);
                    InsertProduct(conn, tx, "p-04", "PH-IP11", "iPhone 11 64GB", "cat-phones", 1900, 0, 2200, 8, now);
                    InsertProduct(conn, tx, "p-05", "TB-IPAD9", "iPad 9 64GB", "cat-tablets", 1500, 0, 1780, 10, now);
                    InsertProduct(conn, tx, "p-06", "TB-SM-T220", "Samsung Tab A7 Lite", "cat-tablets", 650, 0, 800, 15, now);
                    InsertProduct(conn, tx, "p-07", "TB-LN-M10", "Lenovo Tab M10", "cat-tablets", 550, 0, 690, 12, now);
                    InsertProduct(conn, tx, "p-08", "AC-CH-25W", "Charger 25W Fast", "cat-accessories", 35, 0, 65, 100, now);
                    InsertProduct(conn, tx, "p-09", "AC-CB-C", "USB-C Cable 1m", "cat-accessories", 8, 0, 20, 200, now);
                    InsertProduct(conn, tx, "p-10", "AC-HP-BT", "Bluetooth Headset", "cat-accessories", 90, 0, 150, 60, now);
                    InsertProduct(conn, tx, "p-11", "AC-CV-LTH", "Leather Cover", "cat-accessories", 25, 0, 55, 80, now);
                    InsertProduct(conn, tx, "p-12", "SP-BT-IP", "iPhone Battery", "cat-parts", 60, 0, 120, 40, now);
                    InsertProduct(conn, tx, "p-13", "SP-SC-SM", "Samsung Screen", "cat-parts", 180, 0, 280, 25, now);
                    InsertProduct(conn, tx, "p-14", "SP-IC-CH", "Charging IC", "cat-parts", 15, 0, 45, 100, now);

                    // Customers (5)
                    InsertCustomer(conn, tx, "cu-01", "Cash Customer", "", "", 0, 0);
                    InsertCustomer(conn, tx, "cu-02", "Ahmed Al-Fitouri", "0910000001", "Tripoli", 5000, 0);
                    InsertCustomer(conn, tx, "cu-03", "Mohamed Ben Ali", "0910000002", "Tripoli", 3000, 0);
                    InsertCustomer(conn, tx, "cu-04", "Sara Khalil", "0910000003", "Benghazi", 2000, 0);
                    InsertCustomer(conn, tx, "cu-05", "Khaled Misrati", "0910000004", "Misrata", 2000, 0);

                    // Suppliers (3)
                    InsertSupplier(conn, tx, "su-01", "Al-Madar Supplier", "021000001", "Tripoli");
                    InsertSupplier(conn, tx, "su-02", "Libya Phone Est.", "021000002", "Tripoli");
                    InsertSupplier(conn, tx, "su-03", "Tripoli Accessories Co.", "021000003", "Tripoli");

                    // Settings
                    UpsertSetting(conn, tx, "store_name", "Dokkanek");
                    UpsertSetting(conn, tx, "currency", "د.ل");
                    UpsertSetting(conn, tx, "schema_version", "1");

                    tx.Commit();
                }
            }
        }

        /// <summary>Compat with web lib/auth.ts: hex(SHA256("dokkanek:"+password)).</summary>
        public static string HashPassword(string password)
        {
            if (password == null)
            {
                password = string.Empty;
            }
            using (SHA256 sha = SHA256.Create())
            {
                byte[] bytes = Encoding.UTF8.GetBytes("dokkanek:" + password);
                byte[] hash = sha.ComputeHash(bytes);
                StringBuilder sb = new StringBuilder(hash.Length * 2);
                for (int i = 0; i < hash.Length; i++)
                {
                    sb.Append(hash[i].ToString("x2"));
                }
                return sb.ToString();
            }
        }

        private static void InsertUser(SQLiteConnection conn, SQLiteTransaction tx, string id, string name, string username, string hash, string role, string branchId, string now)
        {
            Exec(conn, tx, "INSERT INTO \"User\"(Id, Name, Username, PasswordHash, Role, Active, BranchId, CreatedAt) VALUES(@id,@name,@u,@ph,@role,@active,@branch,@created);",
                P("@id", id), P("@name", name), P("@u", username), P("@ph", hash),
                P("@role", role), P("@active", 1), P("@branch", branchId), P("@created", now));
        }

        private static void InsertProduct(SQLiteConnection conn, SQLiteTransaction tx, string id, string sku, string name, string categoryId, double cost, double usd, double price, double qty, string now)
        {
            Exec(conn, tx, "INSERT INTO \"Product\"(Id, Sku, Name, CategoryId, CostPrice, CostUsd, SalePrice, Quantity, WarehouseId, Barcode, IsFavorite, MinQuantity, Active, LegacyNo, CreatedAt, UpdatedAt) VALUES(@id,@sku,@name,@cat,@cost,@usd,@price,@qty,@wh,@bc,@fav,@min,@active,@legacy,@created,@updated);",
                P("@id", id), P("@sku", sku), P("@name", name), P("@cat", categoryId),
                P("@cost", cost), P("@usd", usd), P("@price", price), P("@qty", qty),
                P("@wh", "wh-main"), P("@bc", sku), P("@fav", 0), P("@min", 5),
                P("@active", 1), P("@legacy", string.Empty), P("@created", now), P("@updated", now));
        }

        private static void InsertCustomer(SQLiteConnection conn, SQLiteTransaction tx, string id, string name, string phone, string address, double limit, double balance)
        {
            Exec(conn, tx, "INSERT INTO \"Customer\"(Id, Name, Phone, Address, CreditLimit, Balance) VALUES(@id,@name,@phone,@addr,@limit,@bal);",
                P("@id", id), P("@name", name), P("@phone", phone), P("@addr", address),
                P("@limit", limit), P("@bal", balance));
        }

        private static void InsertSupplier(SQLiteConnection conn, SQLiteTransaction tx, string id, string name, string phone, string address)
        {
            Exec(conn, tx, "INSERT INTO \"Supplier\"(Id, Name, Phone, Address, Balance) VALUES(@id,@name,@phone,@addr,@bal);",
                P("@id", id), P("@name", name), P("@phone", phone), P("@addr", address), P("@bal", 0d));
        }

        private static void UpsertSetting(SQLiteConnection conn, SQLiteTransaction tx, string key, string value)
        {
            Exec(conn, tx, "INSERT INTO \"Setting\"(Key, Value) VALUES(@k,@v) ON CONFLICT(Key) DO UPDATE SET Value = @v;",
                P("@k", key), P("@v", value));
        }

        private static SQLiteParameter P(string name, object value)
        {
            return new SQLiteParameter(name, value ?? DBNull.Value);
        }

        private static void Exec(SQLiteConnection conn, SQLiteTransaction tx, string sql, params SQLiteParameter[] parameters)
        {
            using (SQLiteCommand cmd = conn.CreateCommand())
            {
                cmd.Transaction = tx;
                cmd.CommandText = sql;
                for (int i = 0; i < parameters.Length; i++)
                {
                    cmd.Parameters.Add(parameters[i]);
                }
                cmd.ExecuteNonQuery();
            }
        }
    }
}
